lib/particle/device.rb

```diff
@@ -58,8 +55,37 @@ def variables
       @attributes[:variables]
     end
 
+    def _product
+      @_product ||= Product.new(@client, product_id)
+    end
+
+    def platform
+      @platform ||= Platform.new(@client, platform_id)
+    end
+
+    def platform_name
+      platform.name
+    end
+
     def product
-      PRODUCT_IDS[product_id]
+      return _product if part_of_product?
+
+      puts <<~BUTT
+        DEPRECATION WARNING:
+        Using Device#product for retrieving the device's platform (i.e. 'Photon', 'Core', etc.) is deprecated; please use Device#platform_name instead (or, use Device#platform to get full platform object).
+        Behavior for Device#product will change in version 2.x.x, such that devices without a user-defined product will soon return nil.
+        (called from: #{caller(1..1).first})
+      BUTT
+
+      platform_name
+    end
+
+    # If the device isn't part of a Product that the user created,
+    # then product_id will return the platform_id. So, conversely,
+    # if product_id is not a platform_id, then it is a part of a
+    # user-created product.
+    def part_of_product?
```

Comment:
Let's name this `dev_kit?` as it indicates that a device is a dev kit used for prototyping vs a device used as part of a product.